packages/aws-cdk/lib/api/hotswap/common.ts

```diff
@@ -18,33 +18,49 @@ export interface HotswapOperation {
    */
   readonly resourceNames: string[];
 
-  apply(sdk: ISDK): Promise<any>;
+  readonly apply: (sdk: ISDK) => Promise<void>;
 }
 
-/**
- * An enum that represents the result of detection whether a given change can be hotswapped.
- */
-export enum ChangeHotswapImpact {
+export interface NonHotswappableChange {
+  readonly hotswappable: false;
+  readonly resourceType: string;
+  readonly rejectedChanges: Array<string>;
+  readonly logicalId: string;
   /**
-   * This result means that the given change cannot be hotswapped,
-   * and requires a full deployment.
+   * Tells the user exactly why this change was deemed non-hotswappable and what its logical ID is.
+   * If not specified, `reason` will be autofilled to state that the properties listed in `rejectedChanges` are not hotswappable.
    */
-  REQUIRES_FULL_DEPLOYMENT = 'requires-full-deployment',
+  readonly reason?: string;
+}
+
+export type ChangeHotswapResult = Array<HotswappableChange | NonHotswappableChange>;
 
+export interface ClassifiedResourceChanges {
+  hotswappableChanges: HotswappableChange[];
+  nonHotswappableChanges: NonHotswappableChange[];
+  metadataChanged: boolean;
+}
+
+export enum HotswapMode {
   /**
-   * This result means that the given change can be safely be ignored when determining
-   * whether the given Stack can be hotswapped or not
-   * (for example, it's a change to the CDKMetadata resource).
+   * Will fall back to CloudFormation when a non-hotswappable change is detected
    */
-  IRRELEVANT = 'irrelevant',
-}
+  CLASSIC = 'classic',
```

Comment:
This doesn't mean much to me. Maybe we could find a more descriptive name for it?
I struggle to come up with one though, and it might be easier if this wouldn't just be a `HotswapMode` but a generic `WatchDeploymentMode`.

Reply:
:+1: how about `FALL_BACK`?